### DISCUSSION TITLE
How to access a returned value from a Nornir task?

### DISCUSSION THREAD

**Discussion by fdervisi** [nornir-automation/nornir#722] [General]
Dear Experts,

How can I access a returned value, from a Nornir task? Here is an example:

**def start_ping_verification()** will issue an other task which should do a full mesh ping. In **def ping_verification(task,linux)** I want to return, all the devices, which failed as a dictionary, but I don't know how I can access ist, because of the **MultiResult** object!?

Can somebody please help me?

```
def start_ping_verification():
    # Initialize Nornir object from config_file
    nr = InitNornir(config_file="config.yaml")

    # Filter for Hosts
    linux = nr.filter(F(groups__contains="linux"))

    # send cisco config via netmiko
    result = linux.run(task=ping_verification,linux=linux)
    

def ping_verification(task,linux):
    my_dict = {task.host:[]}
    for host in linux.inventory.hosts:
        linux_ip = linux.inventory.hosts[host]['ip']
        result = task.run(
            task=netmiko_send_command, command_string=f"ping {linux_ip} -c1"
            )
        if "0% packet loss" in str(result[0]):
            print(f"No Packet Loss from {task.host} to {host}")
            my_dict[task.host].append(host)
        else:
            print(f"*** Error: Packet Loss from {task.host} to {host} ***")

    return(my_dict)
```

**Comment by bytinbit** [comment]
Hi fdervisi, first of all, try to separate the task execution and the evaluation of the task's result. It helps with separation of concerns (i.e. don't do too many different things in one function) and maintaining code. In your case, I'd return just the result and *then* do whatever it must be done with `result`.

However, I've run into similar problems you had, and solved it by wrapping per-host-results into another Result, basically executing per-host-pings as subtasks. Maybe try this:
```Python
def ping_verification(task, linux):
    for host in linux.inventory.hosts:
        linux_ip = linux.inventory.hosts[host]['ip']
        result = task.run(
            task=netmiko_send_command, command_string=f"ping {linux_ip} -c1"
            )
        return Result(host=task.host, result="All pings executed")
```

I find the evaluation of nornir results always a little bit bewildering (What's the structure of the result's data? It always seems to differ!), so I usually set a breakpoint at `result`, look at the variables in memory and see how I can access what I need. Most IDEs will give you a visual representation of what's in memory and allow you to navigate through the data structures. I find that easier than using Python's built-in `pdb` via the command-line.

Just two side notes: 
1. The `nornir_napalm` plugin already has a `napalm_ping` function.
2. `return my_dict` is more Pythonic than `return(my_dict)`, you can skip the brackets.

Hope this helps!

**Reply by fdervisi** [reply]
Thanks for your answer. It still does not work, this is the output:

```
****** result  *********
**'MultiResult' object is not callable**
PING 10.0.5.10 (10.0.5.10) 56(84) bytes of data.
64 bytes from 10.0.5.10: icmp_seq=1 ttl=252 time=315 ms

--- 10.0.5.10 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 315.604/315.604/315.604/0.000 ms
PING 10.2.3.10 (10.2.3.10) 56(84) bytes of data.
64 bytes from 10.2.3.10: icmp_seq=1 ttl=252 time=2.42 ms
```

and this is the modified script:

```
def ping_verification(task,linux):
    my_dict = {task.host:[]}
    for host in linux.inventory.hosts:
        linux_ip = linux.inventory.hosts[host]['ip']
        result = task.run(
            task=netmiko_send_command, command_string=f"ping {linux_ip} -c1"
            )
        if "0% packet loss" in str(result[0]):
            print(f"No Packet Loss from {task.host} to {host}")
            my_dict[task.host].append(host)
        else:
            print(f"*** Error: Packet Loss from {task.host} to {host} ***")

    return result(host=task.host, result="All pings executed")
```

Maybe I miss something?

**Reply by bytinbit** [reply]
Yes - you still evaluate the result in the task itself. Also you return `result`, not a new `Result`-object. Try maybe this (coding from memory and not having your test infrastructure - there may be errors):

```Python
def ping_verification(task, linux):
    for host in linux.inventory.hosts:
        linux_ip = linux.inventory.hosts[host]['ip']
        result = task.run(
            task=netmiko_send_command, command_string=f"ping {linux_ip} -c1"
            )
    return Result(host=task.host, result="All pings executed")

# initialize nornir with the required inventory then:

overall_results = initialized_nornir.run(task=ping_verification, *args)  # args or something like that to pass on your linux variable

# evaluate overall_result, adapt your code to suit the structure, add code as needed and do the evaluation here, adapt as needed:
if "0% packet loss" in str(result[0]):
    print(f"No Packet Loss from {task.host} to {host}")
    my_dict[task.host].append(host)
else:
    print(f"*** Error: Packet Loss from {task.host} to {host} ***")
``` 
Maybe how I've done it [here](https://github.com/INSRapperswil/nuts/blob/677375c1fd388c35fc7ecbe059a5c609e8b509ee/nuts/base_tests/napalm_ping.py#L109) helps you figuring it out better - it uses `napalm_ping` though, but the concept is the same.

Also: what we want is a task that triggers other subtasks per host. We have Task A (`ping_verification`) for all hosts, that triggers Task B (Subtask `netmiko_send_command`) per host.

In the project I've just linked to, [Task A is rather hidden ](https://github.com/INSRapperswil/nuts/blob/677375c1fd388c35fc7ecbe059a5c609e8b509ee/nuts/context.py#L112), while Subtask is just the [link from before](https://github.com/INSRapperswil/nuts/blob/677375c1fd388c35fc7ecbe059a5c609e8b509ee/nuts/base_tests/napalm_ping.py#L109)

Again, hope that helps a bit, good luck!
